libraries/ms-drivers/src/voltage_regulator.c

```diff
@@ -0,0 +1,52 @@
+#include "voltage_regulator.h"
+
+#include <stddef.h>
+
+#include "gpio.h"
+#include "log.h"
+
+#include "interrupt.h"
+#include "soft_timer.h"
+
+static void prv_timer_callback(SoftTimerId timer_id, void *context) {
+  VoltageRegulatorStorage *storage = context;
+  GpioState state = NUM_GPIO_STATES;
+  gpio_get_state(&storage->monitor_pin, &state);
+  if (state == GPIO_STATE_HIGH && !storage->regulator_on) {
+    storage->error_callback(VOLTAGE_REGULATOR_ERROR_ON_WHEN_SHOULD_BE_OFF,
+                            storage->error_callback_context);
+  } else if (state == GPIO_STATE_LOW && storage->regulator_on) {
+    storage->error_callback(VOLTAGE_REGULATOR_ERROR_OFF_WHEN_SHOULD_BE_ON,
+                            storage->error_callback_context);
+  }
```

Comment:
You didn't restart the timer at the end of the callback function, so this will only be called once. Please add another `soft_timer_start_millis` call at the end here.